### DISCUSSION TITLE
MOBIKE related question

### DISCUSSION THREAD

**Discussion by teedu** [strongswan/strongswan#1745] [General]
As I am not sure wether the following is a bug or lack of my knowledge, then I'll write it under discussion.
I have a remote access ikev2 configuration:
```
connections {
  ra-1 {
    local_addrs = y.y.y.46
    send_cert = always
    send_certreq = no
    pools = ra-pool-1
    local {
      auth = pubkey
      certs = blah.pem
      id = %any
    }
    remote {
      eap_id = %any
      auth = eap-radius
    }
    children {
      ra-1 {
        local_ts = 0.0.0.0/0
        rekey_time = 3600
      }
    }
    version = 2
    proposals = aes256gcm16-sha2_384-modp1024, aes256-sha2_256-modp2048
    rekey_time = 28800
    unique = replace
  }
}

pools {
  ra-pool-1 {
    addrs = 10.101.101.0/24
  }
}
```
And ip address related configuration looks like this:
```
lo               UNKNOWN        yy.yy.yy.40/29 
inet       UP             172.20.104.151/31 fe80::6ccc:35ff:fed8:2059/64 
lan      UP             172.20.104.255/31 fe80::7c78:e0ff:fecc:f409/64 
```
MOBIKE capable client connects successfully and everything flows as expected. Until ca 30 seconds when I see in the logs:
```
Jun 20 07:07:58 linux swan: 14[IKE] <ra-1|1878> IKE_SA ra-1[1878] established between y.y.y.46[y.y.y.46]...x.x.x.204[192.168.7.179]
Jun 20 07:07:58 linux swan: 14[IKE] <ra-1|1878> CHILD_SA ra-1{343} established with SPIs cc399764_i 7a01c3a2_o and TS 0.0.0.0/0 === 10.101.101.5/32
Jun 20 07:07:58 linux swan: 14[NET] <ra-1|1878> sending packet: from y.y.y.46[4500] to x.x.x.204[4500] (269 bytes)

Jun 20 07:15:57 linux swan: 06[IKE] <ra-1|1878> old path is not available anymore, try to find another
Jun 20 07:15:57 linux swan: 06[IKE] <ra-1|1878> looking for a route to x.x.x.204 ...
Jun 20 07:15:57 linux swan: 06[IKE] <ra-1|1878> sending address list update using MOBIKE, implicitly requesting an address change
Jun 20 07:15:57 linux swan: 06[ENC] <ra-1|1878> generating INFORMATIONAL request 0 [ ]
Jun 20 07:15:57 linux swan: 06[IKE] <ra-1|1878> checking path 172.20.104.151[4500] - x.x.x.204[4500]
Jun 20 07:15:57 linux swan: 06[NET] <ra-1|1878> sending packet: from 172.20.104.151[4500] to x.x.x.204[4500] (57 bytes)
Jun 20 07:16:00 linux swan: 14[IKE] <ra-1|1878> path probing attempt 1
....
Jun 20 07:16:22 linux swan: 01[IKE] <ra-1|1878> path probing attempt 10
Jun 20 07:16:22 linux swan: 01[IKE] <ra-1|1878> checking path 172.20.104.151[4500] - x.x.x.204[4500]
Jun 20 07:16:22 linux swan: 01[NET] <ra-1|1878> sending packet: from 172.20.104.151[4500] to x.x.x.204[4500] (57 bytes)
Jun 20 07:16:25 linux swan: 11[IKE] <ra-1|1878> giving up after 10 path probings
Jun 20 07:16:25 linux swan: 11[CFG] <ra-1|1878> sending RADIUS Accounting-Request to server 'primary'
Jun 20 07:16:25 linux swan: 11[CFG] <ra-1|1878> received RADIUS Accounting-Response from server 'primary'
Jun 20 07:16:25 linux swan: 11[CFG] <ra-1|1878> lease 10.101.101.5 by 'user' went offline
```
My question is - why path checking (and subsequent packet sending) uses lan interface address when configuration states that it should use address from the local_addrs value (which is on loopback interface)?
I have read swanctl.conf manual and mobike related stuff, but I haven't found anything that could help me further. Please advise!

**Comment by tobiasbrunner** [comment]
> ```
> 
> ```
> 
> 
>     
>   
> 
> My question is - why path checking (and subsequent packet sending) uses lan interface address when configuration states that it should use address from the local_addrs value (which is on loopback interface)?

Because that address or route is apparently gone, so a new path is searched, checking all available local and remote addresses.

**Comment by teedu** [comment]
Unfortunately this is not an issue. Other clients without mobike can send and receive traffic (having IKE SA bound) on that same interface/IP. Which table does strongswan search routes - in all tables (according to their preference) or just 220?

**Reply by tobiasbrunner** [reply]
> Other clients without mobike can send and receive traffic (having IKE SA bound) on that same interface/IP.

Yes, there is a check that disables path checks if MOBIKE is not supported/disabled and a specific local IP is configured. With MOBIKE routes/addresses are checked. So consider what IPs/routes can be found by strongSwan and also why a roam event is triggered in the first place (are interfaces/addresses/routes changed on the system at that time?).

> Which table does strongswan search routes - in all tables (according to their preference) or just 220?

In all BUT 220 (that's one of the reasons strongSwan installs its routes in a separate table, by default, so they can easily be ignored to avoid loops).

**Comment by teedu** [comment]
Thanks Tobias for your pointers. I managed to play a bit with my setup (including strongswan) and the results are as follows:
- My setup has 2 BGP neighbors (from uplink default route is advertised to the server, and from downlink the server gets multiple LAN subnets)
- Strongswan client (using MOBIKE) successfully connects to the server and the tunnel is up. Until there is a change in the routing table (e.g. one of the LAN subnets disappears or (re)appears from different path)
    - I created in the router a route-map to deny specific LAN subnets and could successfully replicate this issue with StrongSwan.
- In the logs:
```
Jun 26 17:49:25 linux swan: 06[IKE] <ra-1|6> old path is not available anymore, try to find another
Jun 26 17:49:25 linux swan: 06[IKE] <ra-1|6> looking for a route to x.x.x.204 ...
Jun 26 17:49:25 linux swan: 06[IKE] <ra-1|6> sending address list update using MOBIKE, implicitly requesting an address change
Jun 26 17:49:25 linux swan: 06[IKE] <ra-1|6> generating INFORMATIONAL request 0 [ ]
Jun 26 17:49:25 linux swan: 06[IKE] <ra-1|6> checking path 172.20.104.151[4500] - x.x.x.204[4500]
Jun 26 17:49:25 linux swan: 06[IKE] <ra-1|6> sending packet: from 172.20.104.151[4500] to x.x.x.204[4500] (57 bytes)
Jun 26 17:49:25 linux swan: 06[IKE] <ra-1|6> path probing attempt 1
``` 
If I disable roaming events in the strongswan.conf, everything works fine.
So, I guess it is still a bug in strongswan code, that it does not obey local_addrs configuration.

My setup has the following:
```
swanctl -S
strongSwan 5.9.9 swanctl

cat /etc/debian_version 
11.6

uname -a
Linux lab-vgw1 5.10.0-20-amd64 #1 SMP Debian 5.10.158-2 (2022-12-13) x86_64 GNU/Linux

```

**Reply by tobiasbrunner** [reply]
> So, I guess it is still a bug in strongswan code, that it does not obey local_addrs configuration.

It's not a bug. As the log says, there is a apparently no path/route to the peer via the current local IP address. At least none that strongSwan recognizes. How do the routing tables look like initially and after this network change?

Note that you could switch to a different route lookup that directly queries the kernel by setting `charon.plugins.kernel-netlink.fwmark` to `!<mark>` where `<mark>` is an arbitrary value.

**Reply by teedu** [reply]
This is my lab setup:
x.x.x.236 = server loopback address
y.y.y.126 = client NAT address

Remote Access client connects successfully:
```
Jun 27 10:23:18 lab-vgw1 blah-swan: 09[IKE] <ra-1|1> IKE_SA ra-1[1] established between x.x.x.236[x.x.x.236]...y.y.y.126[10.0.126.170]
Jun 27 10:23:18 lab-vgw1 blah-swan: 09[IKE] IKE_SA ra-1[1] established between x.x.x.236[x.x.x.236]...y.y.y.126[10.0.126.170]
Jun 27 10:23:18 lab-vgw1 blah-swan: 09[IKE] <ra-1|1> scheduling rekeying in 13330s
Jun 27 10:23:18 lab-vgw1 blah-swan: 09[IKE] scheduling rekeying in 13330s
Jun 27 10:23:18 lab-vgw1 blah-swan: 09[IKE] <ra-1|1> maximum IKE_SA lifetime 14770s
Jun 27 10:23:18 lab-vgw1 blah-swan: 09[IKE] maximum IKE_SA lifetime 14770s
Jun 27 10:23:18 lab-vgw1 blah-swan: 09[CFG] <ra-1|1> selected proposal: ESP:AES_CBC_256/HMAC_SHA1_96/NO_EXT_SEQ
Jun 27 10:23:18 lab-vgw1 blah-swan: 09[CFG] selected proposal: ESP:AES_CBC_256/HMAC_SHA1_96/NO_EXT_SEQ
Jun 27 10:23:18 lab-vgw1 blah-swan: 09[IKE] <ra-1|1> CHILD_SA ra-1{1} established with SPIs cd60c91f_i 7cb0e020_o and TS 0.0.0.0/0 === 10.101.101.1/32
Jun 27 10:23:18 lab-vgw1 blah-swan: 09[IKE] CHILD_SA ra-1{1} established with SPIs cd60c91f_i 7cb0e020_o and TS 0.0.0.0/0 === 10.101.101.1/32
Jun 27 10:23:18 lab-vgw1 blah-swan: 09[CFG] <ra-1|1> sending RADIUS Accounting-Request to server 'primary'
Jun 27 10:23:18 lab-vgw1 blah-swan: 09[CFG] sending RADIUS Accounting-Request to server 'primary'
Jun 27 10:23:18 lab-vgw1 blah-swan: 09[CFG] <ra-1|1> received RADIUS Accounting-Response from server 'primary'
Jun 27 10:23:18 lab-vgw1 blah-swan: 09[CFG] received RADIUS Accounting-Response from server 'primary'
```

```
root@lab-vgw1:/etc# date && ip route show table all | grep -E "table local|default|10.220." | grep -v medium
Tue 27 Jun 2023 10:23:40 AM UTC
default nhid 13 via 172.20.107.0 dev inet proto bgp metric 20 
10.220.1.0/24 nhid 47 via 172.20.107.252 dev l3vpn proto bgp metric 20 
10.220.1.1 nhid 47 via 172.20.107.252 dev l3vpn proto bgp metric 20 
local x.x.x.236 dev lo table local proto kernel scope host src x.x.x.236 
local 172.20.107.1 dev inet table local proto kernel scope host src 172.20.107.1 
local 172.20.107.253 dev l3vpn table local proto kernel scope host src 172.20.107.253 
```

I trigger bgp routing change (where 10.220.1.x becomes unreachable):
```
root@lab-vgw1:/etc# date && ip route show table all | grep -E "table local|default|10.220." | grep -v medium
Tue 27 Jun 2023 10:23:58 AM UTC
default nhid 13 via 172.20.107.0 dev inet proto bgp metric 20 
local x.x.x.236 dev lo table local proto kernel scope host src x.x.x.236 
local 172.20.107.1 dev inet table local proto kernel scope host src 172.20.107.1 
local 172.20.107.253 dev l3vpn table local proto kernel scope host src 172.20.107.253
```

And the log from strongswan:
```
Jun 27 10:23:55 lab-vgw1 blah-swan: 13[IKE] <ra-1|1> old path is not available anymore, try to find another
Jun 27 10:23:55 lab-vgw1 blah-swan: 13[IKE] old path is not available anymore, try to find another
Jun 27 10:23:55 lab-vgw1 blah-swan: 13[IKE] <ra-1|1> looking for a route to y.y.y.126 ...
Jun 27 10:23:55 lab-vgw1 blah-swan: 13[IKE] looking for a route to y.y.y.126 ...
Jun 27 10:23:55 lab-vgw1 blah-swan: 13[IKE] <ra-1|1> sending address list update using MOBIKE, implicitly requesting an address change
Jun 27 10:23:55 lab-vgw1 blah-swan: 13[IKE] sending address list update using MOBIKE, implicitly requesting an address change
Jun 27 10:23:55 lab-vgw1 blah-swan: 13[ENC] <ra-1|1> generating INFORMATIONAL request 0 [ ]
Jun 27 10:23:55 lab-vgw1 blah-swan: 13[ENC] generating INFORMATIONAL request 0 [ ]
Jun 27 10:23:55 lab-vgw1 blah-swan: 13[IKE] <ra-1|1> checking path 172.20.107.1[4500] - y.y.y.126[43818]
Jun 27 10:23:55 lab-vgw1 blah-swan: 13[IKE] checking path 172.20.107.1[4500] - y.y.y.126[43818]
Jun 27 10:23:55 lab-vgw1 blah-swan: 13[NET] <ra-1|1> sending packet: from 172.20.107.1[4500] to y.y.y.126[43818] (57 bytes)
Jun 27 10:23:55 lab-vgw1 blah-swan: 13[NET] sending packet: from 172.20.107.1[4500] to y.y.y.126[43818] (57 bytes)
```

**Reply by tobiasbrunner** [reply]
> server loopback address

What does that mean? loopback how? Installed on `lo`?

> I trigger bgp routing change (where 10.220.1.x becomes unreachable):

How is that IP address relevant? Or is `y.y.y.126` (i.e. the peer's IP that we try to reach) actually `10.220.1.126`? And if it becomes unreachable, what surprises you about what happens next? (i.e. the complaint that the current path/route isn't available anymore.) Or is that a completely unrelated route that just triggers the roaming event and the check of the current path? (In which case there probably never was a path/route to begin with, it only worked because the original inbound packets had that destination IP address and strongSwan didn't have a reason to reevaluate that.)

strongSwan looks for a usable source address to reach the peer. It basically checks all routes for the destination and then tries to determine a source IP for the best matches (either via a preferred source IP or selector set on the route, the interface listed in the route, or via the gateway and other routes). It does try to preserve the current IP but if that's not found via that lookup this won't work. So if your local IP is actually just a random address on `lo`, it will not be found unless it is explicitly listed as preferred source in a route to the peer. Again, you can try the kernel lookup to see if that works better.
